core/src/autogluon/core/trainer/abstract_trainer.py

```diff
@@ -3046,3 +3061,62 @@ def calibrate_model(self, model_name: str = None, lr: float = 0.01, max_iter: in
                 logger.log(15, f'Temperature term found is: {temp_scalar}')
                 model.params_aux["temperature_scalar"] = temp_scalar
                 model.save()
+
+    def calibrate_decision_threshold(self,
+                                     X: pd.DataFrame = None,
+                                     y: np.array = None,
```

Comment:
This is an interesting topic, because I almost feel that Optional is redundant when the default is None, as the Optional nature is self-evident by the default value.

For example, `mypi` and IDEs automatically convert it to the Optional type hint when default is None, meaning from a development and usage perspective on IDEs, it already hints correctly (https://stackoverflow.com/a/66922247).

The question is if we want the extra verbosity in our code. I am open to it, but I feel like it would be distracting more than helpful given that IDEs already auto-convert it, and we use None defaults everywhere, so it would be a ton of changes.

Reply:
mypy explicitly flags if you assign `None` to non-optional

Reply:
https://peps.python.org/pep-0484/#union-types
![image](https://github.com/autogluon/autogluon/assets/10080307/ef7181ca-6b54-4d6d-8840-f19fcd7f3909)

Reply:
Looks like [PEP604](https://peps.python.org/pep-0604/) confirms it's ok to replace `Optional[X]` with `X | None`.

Reply:
Will use `X | None` with `from __future__ import annotations` following [PEP 604](https://peps.python.org/pep-0604/) instead of Optional as the pipe logic is more readable.